### PR TITLE
[containers] Enables container support for scenes that use visual obsvervations

### PR DESCRIPTION
**Pull request by eshvk**
Based on @AdamPalmarUnity's setup. Please take your time to test and try it out.

### CODE REVIEW

File: python/unityagents/environment.py [hunk 1]
```diff
@@ -95,11 +96,21 @@ def __init__(self, file_name, worker_id=0,
...
-                 '--port', str(self.port),
-                 '--seed', str(seed)])
-
+            if docker_training == False:
+                proc1 = subprocess.Popen(
+                    [launch_string,
+                     '--port', str(self.port),
+                     '--seed', str(seed)])
+            else:
+                docker_ls = ("exec xvfb-run --auto-servernum"
```
Comment (mmattar) [inline python/unityagents/environment.py]: A comment explaining docker_ls would be very helpful here, particularly with regards to the dimensions (640x480x24).

File: docs/Using-Docker.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Using Docker For ML-Agents (Experimental)
 
-We currently offer an experimental solution for Windows and Mac users who would like to do training or inference using Docker. This option may be appealing to those who would like to avoid installing Python and TensorFlow themselves. The current setup forces both TensorFlow and Unity to _only_ rely on the CPU for computations. Consequently, our Docker support is limited to environments whose agents **do not** use camera-based visual observations. For example, the [GridWorld](Learning-Environment-Examples.md#gridworld) environment is **not** supported.
+We currently offer a solution for Windows and Mac users who would like to do training or inference using Docker. This option may be appealing to those who would like to avoid installing Python and TensorFlow themselves. The current setup forces both TensorFlow and Unity to _only_ rely on the CPU for computations. Consequently, our Docker simulation [uses the CPU](https://en.wikipedia.org/wiki/Xvfb) to do visual rendering. This means that environments which involve agents using camera-based visual observations might be slower.
```
Comment (mmattar) [inline docs/Using-Docker.md]: - Linking to https://en.wikipedia.org/wiki/Xvfb in "uses the CPU" was a bit counter-intuitive for me.
- The very end of this sentence "might be slower", presumably you mean training and inference speeds. I would clarify that. (Optional: is there anything we can say regarding how much slower we expect it to run?)

File: docs/Using-Docker.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # Using Docker For ML-Agents (Experimental)
```
Comment (mmattar) [inline docs/Using-Docker.md]: - Remove "(Experimental)"? If so, then also remove it from the Repo Homepage (features section).
- (Minor) In this page we used 3D Ball as the running example for container and env names. But now the image "docker_build_settings.png" links to the Gridworld scene. I think we ought to either update the naming conventions or have the image point to 3Dball, just for consistency.

File: docs/Using-Docker.md (hunk 2 above)
Comment (eshvk) [reply docs/Using-Docker.md]: > Linking to https://en.wikipedia.org/wiki/Xvfb in "uses the CPU" was a bit counter-intuitive for me.
`Xvfb` is a virtualized XServer which does not use the GPU, would you prefer if the text hyperlinked said `does not use the GPU` ? 

> The very end of this sentence "might be slower", presumably you mean training and inference speeds. I would clarify that. (Optional: is there anything we can say regarding how much slower we expect it to run?)

I mean training. Although you could very well use the container for inference as well, it seems more designed for a training use case. I will clarify there. 

We are not significantly slower for the environments we support currently. Having said that, we are using the CPU to do computation which is typically done in a GPU. So when we do get to more complex and rich rendering scenes, I expect performance to drop. That phrase was mostly added to ward off potential issues popping up on Github in the mean time.

File: docs/Using-Docker.md (hunk 3 above)
Comment (eshvk) [reply docs/Using-Docker.md]: Both good points.  Will fix.

**Comment by awjuliani** [comment]
Would it be possible to provide a lightweight container for only doing vector observations, and this new one for visual observations? I think if it isn't necessary, we should direct people to use headless mode and the slimmer image.

**Comment by eshvk** [comment]
@awjuliani I am not sure I understand what slimmer or lightweight mean in this context. If you mean that there are more lines in the DockerFile, it is because we are taking content from the [previous base image](https://github.com/docker-library/python/blob/c9a08272dbcbcd8f917ac98f494e6f0e3e7d929e/3.6/jessie/slim/Dockerfile). The only changes from the previous container is that we are using Ubuntu and installing xvfb. 


As far as headless is concerned, someone who wants to use only vector observations can very well build a headless Unity build and use it with our current setup. I am happy to add a note on that.

**Comment by Lakrix** [comment]
When trying out docker image from develop-feature-docker-improvements I get the following error.

Debug info from gdb:

mono_gdb_render_native_backtraces not supported on this platform, unable to find gdb or lldb

=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

Traceback (most recent call last):
  File "python/learn.py", line 62, in <module>
    tc.start_learning()
  File "/execute/python/unitytrainers/trainer_controller.py", line 230, in start_learning
    curr_info = self.env.reset(train_mode=self.fast_simulation)
  File "/execute/python/unityagents/environment.py", line 276, in reset
    return self._get_state()
  File "/execute/python/unityagents/environment.py", line 287, in _get_state
    state_dict, end_of_message = self._get_state_dict()
  File "/execute/python/unityagents/environment.py", line 243, in _get_state_dict
    state = self._recv_bytes().decode('utf-8')
  File "/execute/python/unityagents/environment.py", line 219, in _recv_bytes
    message_length = struct.unpack("I", bytearray(s[:4]))[0]
struct.error: unpack requires a buffer of 4 bytes

I run it on a Mac.

**Comment by eshvk** [comment]
Hi @Lakrix, 
  Can you please create a separate issue to track this? This PR is a work in progress being internally tested:

Some useful information would be: 

1. Have you been able to get the environment working without using docker on the mac?

2. Are you using an `x86_64` linux image? 

3. Did you see any docker build errors when your image is being created?

File: docs/Using-Docker.md (hunk 2 above)
Comment (mmattar) [reply docs/Using-Docker.md]: I think if we're going to link to Xvfb then we should probably introduce it here too.

File: docs/Using-Docker.md [hunk 4]
```diff
@@ -26,13 +27,16 @@ Since Docker typically runs a container sharing a (linux) kernel with the host m
...
-- `Check` the _Headless Mode_ option. (_This is required because the Unity binary will run in a container that does not have graphics drivers installed_.) 
 
-![Build Settings For Docker](images/docker_build_settings.png)
+![Build Settings For Docker](images/docker_build_settings_noheadless.png)
 
 Then click `Build`, pick an environment name (e.g. `3DBall`) and set the output directory to `unity-volume`. After building, ensure that the file `<environment-name>.x86_64` and subdirectory `<environment-name>_Data/` are created under `unity-volume`.
 
+**NOTE** If you are only collecting vector observations from Unity, you can select the `headless` option here:
+
+![Build Settings For Docker](images/docker_build_settings_headless.png)
```
Comment (mmattar) [inline docs/Using-Docker.md]: I don't see a difference between the *_headless.png and *_noheadless.png images, except that headless points to the GridWorld scene, while noheadless points to the 3DBall scene.

File: docs/Using-Docker.md [hunk 5]
```diff
@@ -26,13 +27,16 @@ Since Docker typically runs a container sharing a (linux) kernel with the host m
...
 - Set the _Architecture_ to `x86_64`
-- `Uncheck` the _Development Build_ option
-- `Check` the _Headless Mode_ option. (_This is required because the Unity binary will run in a container that does not have graphics drivers installed_.) 
 
-![Build Settings For Docker](images/docker_build_settings.png)
+![Build Settings For Docker](images/docker_build_settings_noheadless.png)
 
 Then click `Build`, pick an environment name (e.g. `3DBall`) and set the output directory to `unity-volume`. After building, ensure that the file `<environment-name>.x86_64` and subdirectory `<environment-name>_Data/` are created under `unity-volume`.
 
+**NOTE** If you are only collecting vector observations from Unity, you can select the `headless` option here:
```
Comment (mmattar) [inline docs/Using-Docker.md]: Consider including the headless commentary in the list above. That why when a user is following the instructions, they better understand what role the headless option represents.

Comment (mmattar) [reply docs/Using-Docker.md]: "If you are only collecting vector observations from Unity" --> "If the environment does not contains visual observations"

Reason for this change too is that headless presumably is about visual, so if they're using vector and/or text observations then they should be ok.

File: docs/Using-Docker.md (hunk 4 above)
Comment (eshvk) [reply docs/Using-Docker.md]: That was a typo. I think I didn't commit the updated file. The `_headless.png` should be identical to the `_noheadless.png` except for the former having `headless` checked.

File: docs/Using-Docker.md (hunk 5 above)
Comment (eshvk) [reply docs/Using-Docker.md]: So I am going to fold the `NOTE` into the checklist. Let me know if you think the two images for the `headless` and `noheadless` options seem useless now?

**Comment by vincentpierre** [comment]
It works for me feel free, the changes I requested are only around docs

**Comment by eshvk** [comment]
@vincentpierre check new annotated image please.

**Pull request merged**